### PR TITLE
Add Python Mode

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62772 CompositeImplicitAutograd compliance testing
* **#62771 Add Python Mode**

This PR adds a (private) enable_factory_dispatch context manager.
(see torch/utils/_torch_dispatch.py).
enable_factory_dispatch accepts an instance of a `__torch_dispatch__` object
as its argument (call it `ref`). Whenever a factory function gets called
inside of the context manager, it will dispatch to the
`__torch_dispatch__` of the `ref` tensor.

There are quite a few changes that were made to support this.

First, the `concrete_dispatch_fn` in `python_variable.cpp` now accepts
an "optional" `const TorchDispatchOverride*` argument.
TorchDispatchOverride is a struct that holds a tensor that has
`__torch_dispatch__` on it. When the TorchDispatchOverride argument is
non-nullptr, then `concrete_dispatch_fn` always selects the
TorchDispatchOverride's tensor's `__torch_dispatch__` to dispatch on.

Why is the optional argument a const pointer and not `optional<T>`? It
has to do with some library shenanigans. This PR also modifies the
signature of `PyInterpreter::dispatch` to match `concrete_dispatch_fn`,
along with other signatures of dispatch functions.
`PyInterpreter::dispatch` is defined in TensorImpl.h and TensorImpl.h
has no clue what a Tensor is yet, so we use a level of indirection to
get past that. See `Note [TorchDispatchOverride indirection]` for more
details.

Next, when the context manager is alive, it sets some TLS that enables a
new dispatch key (DispatchKey::PythonMode). You can find this logic in
PythonMode.{h, cpp}. Constructing `enable_factory_dispatch` with a `ref`
tensor stores the `ref` tensor in a TLS. I made modifications to
`ATen/ThreadLocalState` to get this working properly. When the context
manager goes out of scope, the dispatch key TLS is removed.

Finally, the last major part of this PR is the DispatchKey::PythonMode
registrations. For every "factory function", we register a boxed kernel
that looks up the TLS, grabs the PyInterpreter from the reference
tensor, and invokes `dispatch` by using the reference tensor as the
TorchDispatchOverride. We define factory function to be all operators
that return a Tensor but accept no Tensors and this is all codegen'ed in
changes in `gen.py`. A sample of the generated file can be found
[here](https://gist.github.com/zou3519/1f28d5532d1fa356b928f88211e0b77c).

There are two more miscellaneous changes:
- `enable_factory_dispatch` respects torch._C._DisableTorchDispatch.
This PR adds the PythonMode key to things that _DisableTorchDispatch
disables.
- internal_new_from_data (torch/csrc/utils/tensor_new.cpp) gets an
exclude guard. `enable_factory_dispatch` currently does not handle
torch.tensor and the exclude guard is to prevent a bug.

Test Plan:
- new tests

Differential Revision: [D30117906](https://our.internmc.facebook.com/intern/diff/D30117906)

### CODE REVIEW

**Comment by zou3519** [comment]
Not ready for review yet

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30117906).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30117906).

**Comment by ezyang** [comment]
I think I mostly have questions about the high level API here, specifically if we should be doing this via a reference object (that has `__torch_dispatch__`, ostensibly the same one that object level invocations of dispatch trigger) or possibly some other mechanism. The main reason I'm concerned about the reference object is because the `__torch_dispatch__` function you call doesn't get the actual concrete reference object that you registered via the mode (remember, it's a classmethod!) This isn't a big deal for true Tensor arguments, as you will get them in the args, but it matters more for mode functions, because the ref tensor isn't a true argument and doesn't get passed in. This means if you have multiple ref tensors with the same classmethod but different ref objects, you can't distinguish between these calls (I know, this isn't really addressed in this PR, but if you're planning on using Python Mode to solve the vmap factory function problem, you're going to have to deal with compositionality).

I do agree that reusing something like `__torch_dispatch__` API for both factory functions and non-factory functions seems appropriate.

https://numpy.org/neps/nep-0037-array-module.html also seems relevant.

**Comment by zou3519** [comment]
Discussion with Ed offline yielded the following insights:
1. The user-facing API should pass a  __torch_dispatch__ function as well as a class
2. Python mode should not be limited to factory functions. E.g. for FiniteTensor, if people add two LinearOperator tensors, we want that to dispatch to the python mode
3. Pre-pending the current mode to the list of things to consider for torch-func dispatching seems reasonable
4. Nested mode: nested mode keys should be pushed onto a stack (and that stack should get pre-pended to list of things to consider for dispatch)
5. Priority: (not going to be handled in this PR) If we nest multiple of the same class in torch_dispatch, then we will want a way to specify the priority. Ed has a crazy idea where we generate classes on the fly if we end up nesting multiple of the same class (or at least, using that as the mental model)

Alban had a concern about running out of dispatch keys. We should be able to delete FuncTorch Python soon (we've migrated to core's Python key, but have been leaving FuncTorchPython if we need to go back...)

**Comment by zou3519** [comment]
Superceded by https://github.com/pytorch/pytorch/pull/63496

File: tools/codegen/gen.py [hunk 1]
```diff
@@ -522,6 +522,24 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
         else:
             assert_never(self.target)
 
+# Generates RegisterPythonMode.cpp, a series of kernels which provide
+# overrides of all factory functions for PythonMode. A factory function is defined
+# as an operator that returns a Tensor but accepts no Tensor arguments.
+@dataclass(frozen=True)
```
[the remark this answers is not in the document]
Comment (zou3519) [reply tools/codegen/gen.py:528]: @anjali411 please review https://github.com/pytorch/pytorch/pull/63496, it is the updated version of this PR

**Pull request closed without merging**